// tsdk/lib/sync-files.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.syncSharedFiles = exports.syncAddtionShareFiles = exports.syncAPIConf = exports.syncEntityFiles = exports.syncExtFiles = exports.copySDK = exports.copyShared = exports.copySnippet = exports.addDepsIfNone = exports.copyTsdkrc = exports.syncFiles = void 0;
const glob = require("fast-glob");
const fs_extra_1 = __importDefault(require("fs-extra"));
const path_1 = __importDefault(require("path"));
const config_1 = require("./config");
const delete_files_1 = require("./delete-files");
const symbols_1 = __importDefault(require("./symbols"));
const transform_import_path_1 = require("./transform-import-path");
function syncFiles(noOverwrite = false) {
    return __awaiter(this, void 0, void 0, function* () {
        yield copySDK(noOverwrite);
        yield (0, config_1.parseDeps)();
        yield (0, delete_files_1.deleteFilesBeforeSync)();
        yield syncAddtionShareFiles();
        yield syncAPIConf();
        yield syncEntityFiles();
        yield syncSharedFiles();
    });
}
exports.syncFiles = syncFiles;
function copyTsdkrc() {
    return __awaiter(this, void 0, void 0, function* () {
        // copy .tsdkrc and remove packages/fe-sdk .tsdkrc
        yield fs_extra_1.default.copy(path_1.default.join(__dirname, '../fe-sdk-template', './config/.tsdkrc.json'), path_1.default.join(process.cwd(), '.tsdkrc.json'), { overwrite: false });
    });
}
exports.copyTsdkrc = copyTsdkrc;
function addDepsIfNone() {
    return __awaiter(this, void 0, void 0, function* () {
        // check if have `zod` as dependencies
        // if don't have, add `zod` to dependencies
        const pkgPath = path_1.default.resolve(process.cwd(), 'package.json');
        const content = yield fs_extra_1.default.readFile(pkgPath, 'utf8');
        const contentJSON = JSON.parse(content);
        yield Promise.all([
            ['zod', '^3.19.1'],
            ['change-case', '^4.1.2'],
        ].map(([i, version]) => __awaiter(this, void 0, void 0, function* () {
            if (!contentJSON.dependencies[i]) {
                contentJSON.dependencies[i] = version;
                yield fs_extra_1.default.writeFile(pkgPath, JSON.stringify(contentJSON, null, 2));
                console.log('');
                console.log(symbols_1.default.warning, `\`tsdk\` depends on \`${i}\` for validate schema, so auto add \`${i}\` as dependencies`);
                console.log(symbols_1.default.info, 'You can run `npm install` or `yarn` to install new dependencies');
                console.log('');
            }
            return 1;
        })));
    });
}
exports.addDepsIfNone = addDepsIfNone;
function copySnippet() {
    return __awaiter(this, void 0, void 0, function* () {
        yield fs_extra_1.default.copy(path_1.default.join(__dirname, '../fe-sdk-template', './config/.vscode'), path_1.default.resolve(process.cwd(), config_1.config.monorepoRoot || './', '.vscode'), { overwrite: false });
    });
}
exports.copySnippet = copySnippet;
function copyShared() {
    return __awaiter(this, void 0, void 0, function* () {
        yield fs_extra_1.default.copy(path_1.default.join(__dirname, '../fe-sdk-template', './src/shared/'), path_1.default.join(process.cwd(), config_1.config.baseDir, 'shared'), { overwrite: false });
    });
}
exports.copyShared = copyShared;
const defaultPackageScope = '@SCOPE-NAME';
const defaultPackagePrefix = `${defaultPackageScope}:registry`;
function reconfigPkg() {
    return __awaiter(this, void 0, void 0, function* () {
        // rename package name
        const pkgPath = path_1.default.resolve(process.cwd(), config_1.config.packageDir, config_1.packageFolder, 'package.json');
        const docPkgPath = path_1.default.resolve(process.cwd(), config_1.config.packageDir, config_1.packageFolder, 'website', 'package.json');
        const [content, docContent] = yield Promise.all([
            fs_extra_1.default.readFile(pkgPath, 'utf-8'),
            fs_extra_1.default.readFile(docPkgPath, 'utf-8'),
        ]);
        const pkgContent = JSON.parse(content);
        const docPkgContent = JSON.parse(docContent);
        pkgContent.name = config_1.config.packageName;
        docPkgContent.name = `${config_1.config.packageName}-docs`;
        const scope = config_1.config.packageName.split('/')[0];
        if (!scope[0].startsWith('@') || scope.length <= 1) {
            console.log(symbols_1.default.info, `Add scope name in \`.tsdkrc\` \`packageName\` field.`);
        }
        if (scope === defaultPackageScope) {
            console.log(symbols_1.default.warning, `Update package scope name in \`.tsdkrc\` \`packageName\` field.`);
        }
        if (pkgContent.publishConfig && pkgContent.publishConfig[defaultPackagePrefix]) {
            pkgContent.publishConfig[`${scope}:registry`] = pkgContent.publishConfig[defaultPackagePrefix];
            delete pkgContent.publishConfig[defaultPackagePrefix];
        }
        yield Promise.all([
            fs_extra_1.default.writeFile(pkgPath, JSON.stringify(pkgContent, null, 2)),
            fs_extra_1.default.writeFile(docPkgPath, JSON.stringify(docPkgContent, null, 2)),
        ]);
        yield Promise.all([copyShared(), copySnippet()]);
        const content2 = yield fs_extra_1.default.readFile('./package.json', 'utf-8');
        const pkgJSON = JSON.parse(content2);
        pkgJSON.scripts = Object.assign(Object.assign({}, (pkgJSON.scripts || {})), { 'sync-sdk': 'npx tsdk --sync' });
        yield fs_extra_1.default.writeFile('./package.json', JSON.stringify(pkgJSON, null, 2));
    });
}
function copySDK(noOverwrite) {
    return __awaiter(this, void 0, void 0, function* () {
        console.log(symbols_1.default.bullet, `init ${config_1.ensureDir}`);
        if (!config_1.isCurrentConfigExist) {
            yield copyTsdkrc();
        }
        const existPath = path_1.default.resolve(process.cwd(), config_1.config.packageDir, config_1.packageFolder, `package.json`);
        const isExist = yield fs_extra_1.default.pathExists(existPath);
        if (isExist && noOverwrite) {
            yield reconfigPkg();
            console.log(symbols_1.default.info, `skip init sdk: \`${path_1.default.resolve(process.cwd(), config_1.config.packageDir, config_1.packageFolder)}\` already exist`);
            return;
        }
        yield fs_extra_1.default.ensureDir(config_1.ensureDir);
        console.log(symbols_1.default.success, `mkdir -p ${config_1.ensureDir}`);
        yield fs_extra_1.default.copy(path_1.default.join(__dirname, '../fe-sdk-template'), path_1.default.resolve(process.cwd(), config_1.config.packageDir, config_1.packageFolder), { overwrite: true });
        yield reconfigPkg();
        yield Promise.all(['config'].map((folder) => {
            return fs_extra_1.default.remove(path_1.default.join(config_1.ensureDir, folder));
        }));
        console.log(symbols_1.default.success, `init ${config_1.ensureDir}`);
    });
}
exports.copySDK = copySDK;
/** sync files base extension config */
function syncExtFiles(ext, isEntity = false) {
    return __awaiter(this, void 0, void 0, function* () {
        console.log(symbols_1.default.bullet, `sync *.${ext}.ts files`);
        const pattern = path_1.default
            .join(`${path_1.default.join(...config_1.config.baseDir.split('/'))}`, `**`, `*.${ext}.ts`)
            .replace(/\\/g, '/');
        const files = yield glob(pattern);
        files.sort();
        const indexContentMap = {};
        yield Promise.all(files.map((file, idx) => __awaiter(this, void 0, void 0, function* () {
            const filePath = path_1.default.join(config_1.ensureDir, file.replace(`${config_1.config.baseDir}/`, 'src/'));
            const content = yield (0, transform_import_path_1.transformImportPath)(file, isEntity);
            yield fs_extra_1.default.ensureDir(path_1.default.dirname(filePath));
            let fromPath = path_1.default.relative(`${config_1.ensureDir}/src/`.replace(/\\/g, '/'), filePath.replace('.ts', ''));
            fromPath = path_1.default.normalize(fromPath);
            fromPath = fromPath.startsWith('.') ? fromPath : './' + fromPath;
            indexContentMap[file] = `export * from '${fromPath.replace(/\\/g, '/')}';\n`;
            return fs_extra_1.default.writeFile(filePath, content);
        })));
        const indexContent = files.length > 0 ? files.map((file) => indexContentMap[file]).join('') : (0, config_1.getDefaultContent)();
        yield fs_extra_1.default.writeFile(path_1.default.join(config_1.ensureDir, `src/${ext}-refs.ts`), `${config_1.comment}${indexContent}`);
    });
}
exports.syncExtFiles = syncExtFiles;
/** sync entity files  */
function syncEntityFiles() {
    return __awaiter(this, void 0, void 0, function* () {
        return syncExtFiles(config_1.config.entityExt, true);
    });
}
exports.syncEntityFiles = syncEntityFiles;
/** sync apiconf files */
function syncAPIConf() {
    return __awaiter(this, void 0, void 0, function* () {
        return syncExtFiles(config_1.config.apiconfExt);
    });
}
exports.syncAPIConf = syncAPIConf;
/** sync apiconf files */
function syncAddtionShareFiles() {
    return __awaiter(this, void 0, void 0, function* () {
        return syncExtFiles(config_1.config.shareExt || 'shared');
    });
}
exports.syncAddtionShareFiles = syncAddtionShareFiles;
/** sync shared files */
function syncSharedFiles() {
    return __awaiter(this, void 0, void 0, function* () {
        console.log(symbols_1.default.bullet, `sync shared files`);
        const files = yield glob(config_1.config.sharedDirs.map((i) => path_1.default.join(i, `**/*.ts`).replace(/\\/g, '/')));
        files.sort();
        const indexContentMap = {};
        yield Promise.all(files.map((file) => __awaiter(this, void 0, void 0, function* () {
            const filePath = path_1.default.join(config_1.ensureDir, file.replace(`${config_1.config.baseDir}/`, 'src/'));
            const content = yield (0, transform_import_path_1.transformImportPath)(file);
            yield fs_extra_1.default.ensureDir(path_1.default.dirname(filePath));
            let fromPath = path_1.default.relative(`${config_1.ensureDir}/src/`.replace(/\\/g, '/'), filePath.replace('.ts', ''));
            fromPath = path_1.default.normalize(fromPath);
            fromPath = fromPath.startsWith('.') ? fromPath : './' + fromPath;
            if (fromPath.indexOf('tsdk-types') < 0) {
                indexContentMap[file] = `export * from '${fromPath.replace(/\\/g, '/')}';\n`;
            }
            return fs_extra_1.default.writeFile(filePath, content);
        })));
        const indexContent = files.map((file) => indexContentMap[file]).join('\n');
        yield fs_extra_1.default.writeFile(path_1.default.join(config_1.ensureDir, `src`, `shared-refs.ts`), `${config_1.comment}${indexContent}`);
        console.log(symbols_1.default.success, `sync shared files`);
    });
}
exports.syncSharedFiles = syncSharedFiles;
